### PR TITLE
[MRG] provide "protocol" tests for `Index`, `CollectionManifest`, and `LCA_Database` classes

### LINKED ISSUE #1949: `sourmash index` will soon report the wrong number of nodes
In #1936 I fix `len(SBT)` to report the number of signatures, which is what `Index` classes should do.

but now in `sourmash index` output I get:
```
loaded 317542 sigs; saving SBT under "gtdb-rs207.genomic.k51.sbt.zip"
329801 of 317542 nodes saved
```
...because the second number on the second line is calculated using `__len__` on SBT. I can't win! 😆

### PR DESCRIPTION
**Pull request by ctb**
Use pytest fixtures and parametrized tests to start building standardized test suites for the `Index`, `CollectionManifest`, and `LCA_Database` classes.

This PR creates three new test files, `test_index_protocol.py`, `test_manifest_protocol.py`, and `test_lca_db_protocol.py`, that use fixtures and parametrized tests to evaluate the various `Index`, `CollectionManifest`, and `LCA_Database` classes. Part of the motivation is to provide a clearer and more uniform test suite for the various `Sqlite*` classes being introduced in #1808 and #1933.

While `Index` had a fairly well defined API already, this PR ends up defining (and refactoring) the `CollectionManifest` and `LCA_Database` APIs to better test and support core APIs moving forward.
Both `LCA_Database` and `CollectionManifest` received a lot of refactoring as a result, but no functionality was changed.

In addition to creating these new tests, this PR fixes a few problems caught by the new tests:
* fixes `len(...)` on `SBT` to only count the number of leaves, not the number of internal nodes;
* adding a test for `sourmash lca rankinfo --minimum-num`;
* fixes a column-off-by-one bug in the LCA tests/GTDB spreadsheet;
* implements `__len__` for `LazyLoadedIndex`, and adds some correctness checks to constructor;

Remaining issues:
- [x] `RevIndex` doesn't obey the full `Index` API; punted to https://github.com/sourmash-bio/sourmash/issues/1939
- [x] might want to clean up the basic lca_db and manifest creation code in those protocol tests 🤔 

TODO:
- [x] evaluate the sqlite manifest API in light of the [`sigs-to-manifest` script](https://github.com/sourmash-bio/database-examples/blob/main/sigs-to-manifest.py) and the updating code there;

Creates problem, and then fixes it;
Fixes https://github.com/sourmash-bio/sourmash/issues/1949

### CODE REVIEW

**Comment by ctb** [comment]
@sourmash-bio/devs curious if there is a better approach within the pytest world.

**Comment by ctb** [comment]
Ready for review & merge, @sourmash-bio/devs !

File: tests/test_index.py [hunk 1]
```diff
@@ -2456,6 +2343,24 @@ def test_lazy_loaded_index_3_find(runtmp):
...
 
+def test_lazy_loaded_index_4_nofile(runtmp):
+    # test check for filename must exist
+    with pytest.raises(ValueError) as exc:
+        index.LazyLoadedIndex(runtmp.output('xyz'), True)
+
+    assert "must exist when creating" in str(exc)
+
+
+def test_lazy_loaded_index_4_noanifest(runtmp):
```
Comment (bluegenes) [inline tests/test_index.py]:
```suggestion
def test_lazy_loaded_index_4_nomanifest(runtmp):
```

**Review by bluegenes — APPROVED** [review]
this lgtm! ...but I can't speak to whether there are any better pytest strategies 🤷🏻‍♀️

File: tests/test_index.py (hunk 1 above)
Comment (ctb) [reply tests/test_index.py]: fixed in d69596d6!
